File: PerformanceStudio/CSharp/PerformanceStudio/TreeViewUtility.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Linq;
using System.Text;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media.Imaging;
using devDept.Eyeshot;
using devDept.Graphics;
using devDept.Eyeshot.Entities;

namespace PerformanceStudio
{
    /// <summary>
    /// Utility class to browse assembly structure. The Tag represents the currently examined BlockReference
    /// and is used later for Screen->Tree and Tree->Screen selection.
    /// </summary>
    public static class TreeViewUtility
    {
        /// <summary>
        /// Recursive function to populate the tree. 
        /// </summary>
        /// <param name="tv">The treeView control</param>
        /// <param name="entList">The entity list</param>
        /// <param name="blocks">The block collection</param>
        /// <param name="parentNode">The parent node. Can be null for root level nodes.</param>
        public static void PopulateTree(TreeView tv, List<Entity> entList, BlockKeyedCollection blocks, TreeNode parentNode = null)
        {
            ItemCollection nodes;
            if (parentNode == null)
            {
                tv.Items.Clear();
                nodes = tv.Items;
            }
            else
            {
                nodes = parentNode.Items;
            }

            for (int i = 0; i < entList.Count; i++)
            {
                Entity ent = entList[i];
                if (ent is BlockReference)
                {
                    Block child;
                    string blockName = ((BlockReference)ent).BlockName;

                    if (blocks.TryGetValue(blockName, out child))
                    {
                        TreeNode parentTn = new TreeNode(parentNode, GetNodeName(blockName, i), false);
                        parentTn.Tag = ent;

                        nodes.Add(parentTn);
                        PopulateTree(tv, child.Entities, blocks, parentTn);
                    }
                }
                else
                {
                    string type = ent.GetType().ToString().Split('.').LastOrDefault();
                    var node = new TreeNode(parentNode, GetNodeName(type, i), true);
                    node.Tag = ent;
                    nodes.Add(node);
                }
            }
        }

        private static string GetNodeName(string name, int index)
        {
            return String.Format("{0}", name, index);
        }

        /// <summary>
        /// Clear selection for entities to avoid problems with the Tree->Screen selection
        /// </summary>
        /// <param name="vl">The Model control</param>
        /// <param name="rootLevel">When true the CurrentBlockReference is set to null (Go back to the root level of the assembly)</param>
        public static void CleanCurrent(Model vl, bool rootLevel = true)
        {
            vl.Entities.ClearSelection();

            if (rootLevel && vl.Entities.CurrentBlockReference != null)
                vl.Entities.SetCurrent(null);
        }

        /// <summary>
        /// Clear selection recursively for inner nodes.
        /// </summary>
        /// <param name="blocks">The blocks collection</param>
        /// <param name="parentBr">The parent block reference</param>
        public static void CleanCurrentNodes(BlockKeyedCollection blocks, BlockReference parentBr)
        {
            Block toClean;

            if (blocks.TryGetValue(parentBr.BlockName, out toClean))
            {
                for (int i = 0; i < toClean.Entities.Count; i++)
                {
                    if (toClean.Entities[i] is BlockReference)
                    {
                        toClean.Entities[i].Selected = false;
                        CleanCurrentNodes(blocks, (BlockReference)toClean.Entities[i]);
                    }
                }
            }
        }

        /// <summary>
        /// Deletes the selected tree node and all the others nodes that are linked to the same entity instance.
        /// </summary>
        /// <param name="tv">The TreeView control</param>
        /// <param name="vl">The Model control</param>
        public static void DeleteSelectedNode(TreeView tv, Model vl)
        {
            if (tv.SelectedItem != null)
            {
                Entity deletedEntity = ((TreeNode)tv.SelectedItem).Tag as Entity;
                DeleteNodes(deletedEntity, tv.Items);
                CleanCurrent(vl);
            }
        }

        /// <summary>
        /// Deletes all the tree nodes that are referring to the same entity instance
        /// </summary>
        /// <param name="entity">The entity instance.</param>
        /// <param name="nodes">The TreeNode collection</param>
        private static void DeleteNodes(Entity entity, ItemCollection nodes)
        {
            int count = nodes.Count;
            while (count > 0)
            {
                count--;
                TreeNode node = (TreeNode)nodes[count];
                if (ReferenceEquals(entity, node.Tag))
                {
                    node.Remove();
                    count = -1;
                }
                else
                {
                    DeleteNodes(entity, node.Items);
                }
            }
        }

        /// <summary>
        /// Tree->Screen Selection. If the viewport entities are selected, they get marked as selected straight away.
        /// To check we are considering the correct entities we use the Entity stored in the Tag property of the TreeView
        /// </summary>
        /// <returns>The selected item.</returns>
        /// <param name="tv">The TreeView control</param>
        /// <param name="vl">The Model control</param>        
        public static Model.SelectedItem SynchTreeSelection(TreeView tv, Model vl)
        {
            // Fill a stack of entities and blockreferences starting from the node tags.
            Stack<BlockReference> parents = new Stack<BlockReference>();

            TreeNode node = (TreeNode)tv.SelectedItem;

            if (node != null)
            {
                Entity entity = node.Tag as Entity;

                node = node.ParentNode;

                while (node != null)
                {
                    var ent = node.Tag as Entity;
                    if (ent != null)

                        parents.Push((BlockReference)ent);

                    node = node.ParentNode;
                }

                //tv.HideSelection = false;

                // The top most parent is the root Blockreference: must reverse the order, creating a new Stack
                var selItem = new Model.SelectedItem(new Stack<BlockReference>(parents), entity);

                // Selects the item
                selItem.Select(vl, true);

                return selItem;
            }
            return null;
        }

        /// <summary>
        /// Screen->Tree Selection.
        /// </summary>
        /// <param name="tv">The TreeView control</param>
        /// <param name="blockReferences">The BlockReference stack</param>
        /// <param name="selectedEntity">The selected entity inside a block reference. Can be null when we click on a BlockReference.</param>
        public static void SynchScreenSelection(TreeView tv, Stack<BlockReference> blockReferences, Model.SelectedItem selectedEntity)
        {
            if (tv.SelectedItem != null)
            {
                if (tv.SelectedItem is TreeViewItem)
                    ((TreeViewItem)tv.SelectedItem).IsSelected = false;
                else
                    ((TreeNode)tv.SelectedItem).IsSelected = false;
            }

            CollapseAll(tv);

            if (selectedEntity != null && selectedEntity.Parents.Count > 0)
            {
                //// Add the parents of the selectedEntity to the BlockReferences stack

                // Reverse the stack so the one on top is the one at the root of the hierarchy
                var parentsReversed = selectedEntity.Parents.Reverse();

                var cumulativeStack = new Stack<BlockReference>(blockReferences);

                foreach (var br in parentsReversed)
                {
                    cumulativeStack.Push(br);
                }

                // Create a new stack with the reversed order so the one on top is the root.
                blockReferences = new Stack<BlockReference>(cumulativeStack);
            }

            SearchNodeInTree(tv, blockReferences, selectedEntity);
        }


        /// <summary>
        /// Screen->Tree Selection. To check we are considering the correct entities, we use the Entity stored in the Tag property of the TreeView.
        /// </summary>
        /// <param name="tv">The TreeView control</param>
        /// <param name="blockReferences">The block reference stack</param>
        /// <param name="selectedEntity">The selected entity inside a block reference. Can be null when we click on a BlockReference.</param>
        /// <param name="parentTn">The parent TreeNode for searching inside its nodes. Can be null.</param>
        public static void SearchNodeInTree(TreeView tv, Stack<BlockReference> blockReferences, Model.SelectedItem selectedEntity, TreeNode parentTn = null)
        {
            if (blockReferences.Count == 0 && selectedEntity == null)
                return;

            ItemCollection tnc = tv.Items;
            if (parentTn != null)
                tnc = parentTn.Items;

            if (blockReferences.Count > 0)
            {
                // Nested BlockReferences

                BlockReference br = blockReferences.Pop();

                foreach (TreeNode tn in tnc)
                {
                    if (ReferenceEquals(br, tn.Tag))
                    {
                        if (blockReferences.Count > 0)
                        {
                            tn.IsExpanded = true;
                            SearchNodeInTree(tv, blockReferences, selectedEntity, tn);
                        }
                        else
                        {
                            if (selectedEntity != null)
                            {
                                foreach (TreeNode childNode in tn.Items)
                                {
                                    if (ReferenceEquals(selectedEntity.Item, childNode.Tag))
                                    {
                                        if (childNode.ParentNode != null) childNode.ParentNode.IsExpanded = true;
                                        childNode.IsSelected = true;
                                        break;
                                    }
                                }
                            }
                            else
                            {
                                tn.IsSelected = true;
                            }
                        }

                        return;
                    }
                }
            }
            else
            {
                // Root level

                if (selectedEntity != null)
                {
                    foreach (TreeNode childNode in tnc)
                    {
                        if (ReferenceEquals(selectedEntity.Item, childNode.Tag))
                        {
                            childNode.IsSelected = true;
                            break;
                        }
                    }
                }

            }
        }

        public static void CollapseAll(TreeView tv)
        {
            foreach (TreeNode i in tv.Items)
            {
                i.IsExpanded = false;
            }
        }

    }

    /// <summary>
    /// In the XAML markup, I have specified a HierarchicalDataTemplate for the ItemTemplate of the TreeView.
    /// This class represent the ViewModel for TreeView's Items.
    /// </summary>
    public class TreeNode : FrameworkElement
    {
        public TreeNode(TreeNode parent)
        {
            Items = new TreeView().Items;
            ParentNode = parent;
        }

        public TreeNode(TreeNode parent, string text, bool isPart) : this(parent)
        {
            Text = text;
            if (isPart)
                Icon = new BitmapImage(GetUriFromResource("part_icon.png"));
            else
                Icon = new BitmapImage(GetUriFromResource("component_icon.png"));
        }

        public string Text { get; set; }

        public BitmapImage Icon { get; private set; }

        private static Uri GetUriFromResource(string resourceFilename)
        {
            return new Uri(@"pack://application:,,,/Resources/" + resourceFilename);
        }

        public TreeNode ParentNode { get; set; }

        public ItemCollection Items { get; set; }

        public static readonly DependencyProperty IsSelectedProperty = DependencyProperty.Register(
            "IsSelected", typeof(bool), typeof(TreeNode), new PropertyMetadata(default(bool)));

        public bool IsSelected
        {
            get { return (bool)GetValue(IsSelectedProperty); }
            set { SetValue(IsSelectedProperty, value); }
        }

        public static readonly DependencyProperty IsExpandedProperty = DependencyProperty.Register(
            "IsExpanded", typeof(bool), typeof(TreeNode), new PropertyMetadata(default(bool)));

        public bool IsExpanded
        {
            get { return (bool)GetValue(IsExpandedProperty); }
            set { SetValue(IsExpandedProperty, value); }
        }

        public override string ToString()
        {
            return Text;
        }

        public void Remove()
        {
            while (Items.Count > 0)
            {
                ((TreeNode)Items[0]).Remove();
            }

            if (ParentNode != null)
                ParentNode.Items.Remove(this);
        }
    }

}

File: PerformanceStudio/Visual Basic/PerformanceStudio/TreeViewUtility.vb
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.ComponentModel
Imports System.Linq
Imports System.Text
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Media.Imaging
Imports devDept.Eyeshot
Imports devDept.Graphics
Imports devDept.Eyeshot.Entities

Module TreeViewUtility
        Sub PopulateTree(ByVal tv As TreeView, ByVal entList As List(Of Entity), ByVal blocks As BlockKeyedCollection, ByVal Optional parentNode As TreeNode = Nothing)
            Dim nodes As ItemCollection

            If parentNode Is Nothing Then
                tv.Items.Clear()
                nodes = tv.Items
            Else
                nodes = parentNode.Items
            End If

            For i As Integer = 0 To entList.Count - 1
                Dim ent As Entity = entList(i)

                If TypeOf ent Is BlockReference Then
                    Dim child As Block
                    Dim blockName As String = (CType(ent, BlockReference)).BlockName

                    If blocks.TryGetValue(blockName, child) Then
                        Dim parentTn As TreeNode = New TreeNode(parentNode, GetNodeName(blockName, i), False)
                        parentTn.Tag = ent
                        nodes.Add(parentTn)
                        PopulateTree(tv, child.Entities, blocks, parentTn)
                    End If
                Else
                    Dim type As String = ent.[GetType]().ToString().Split("."c).LastOrDefault()
                    Dim node = New TreeNode(parentNode, GetNodeName(type, i), True)
                    node.Tag = ent
                    nodes.Add(node)
                End If
            Next
        End Sub

    Private Function GetNodeName(ByVal name As String, ByVal index As Integer) As String
        Return String.Format("{0}", name)
    End Function

    Sub CleanCurrent(ByVal vl As Model, ByVal Optional rootLevel As Boolean = True)
            vl.Entities.ClearSelection()
            If rootLevel AndAlso vl.Entities.CurrentBlockReference IsNot Nothing Then vl.Entities.SetCurrent(Nothing)
        End Sub

        Sub CleanCurrentNodes(ByVal blocks As BlockKeyedCollection, ByVal parentBr As BlockReference)
            Dim toClean As Block

            If blocks.TryGetValue(parentBr.BlockName, toClean) Then

                For i As Integer = 0 To toClean.Entities.Count - 1

                    If TypeOf toClean.Entities(i) Is BlockReference Then
                        toClean.Entities(i).Selected = False
                        CleanCurrentNodes(blocks, CType(toClean.Entities(i), BlockReference))
                    End If
                Next
            End If
        End Sub

        Sub DeleteSelectedNode(ByVal tv As TreeView, ByVal vl As Model)
            If tv.SelectedItem IsNot Nothing Then
                Dim deletedEntity As Entity = TryCast((CType(tv.SelectedItem, TreeNode)).Tag, Entity)
                DeleteNodes(deletedEntity, tv.Items)
                CleanCurrent(vl)
            End If
        End Sub

        Private Sub DeleteNodes(ByVal entity As Entity, ByVal nodes As ItemCollection)
            Dim count As Integer = nodes.Count

            While count > 0
                count -= 1
                Dim node As TreeNode = CType(nodes(count), TreeNode)

                If ReferenceEquals(entity, node.Tag) Then
                    node.Remove()
                    count = -1
                Else
                    DeleteNodes(entity, node.Items)
                End If
            End While
        End Sub

        Function SynchTreeSelection(ByVal tv As TreeView, ByVal vl As Model) As Model.SelectedItem
            Dim parents As Stack(Of BlockReference) = New Stack(Of BlockReference)()
            Dim node As TreeNode = CType(tv.SelectedItem, TreeNode)

            If node IsNot Nothing Then
                Dim entity As Entity = TryCast(node.Tag, Entity)
                node = node.ParentNode

                While node IsNot Nothing
                    Dim ent = TryCast(node.Tag, Entity)
                    If ent IsNot Nothing Then parents.Push(CType(ent, BlockReference))
                    node = node.ParentNode
                End While

                Dim selItem = New Model.SelectedItem(New Stack(Of BlockReference)(parents), entity)
                selItem.[Select](vl, True)
                Return selItem
            End If

            Return Nothing
        End Function

        Sub SynchScreenSelection(ByVal tv As TreeView, ByVal blockReferences As Stack(Of BlockReference), ByVal selectedEntity As Model.SelectedItem)
            If tv.SelectedItem IsNot Nothing Then

                If TypeOf tv.SelectedItem Is TreeViewItem Then
                    CType(tv.SelectedItem, TreeViewItem).IsSelected = False
                Else
                    CType(tv.SelectedItem, TreeNode).IsSelected = False
                End If
            End If

            CollapseAll(tv)

            If selectedEntity IsNot Nothing AndAlso selectedEntity.Parents.Count > 0 Then
                Dim parentsReversed = selectedEntity.Parents.Reverse()
                Dim cumulativeStack = New Stack(Of BlockReference)(blockReferences)

                For Each br In parentsReversed
                    cumulativeStack.Push(br)
                Next

                blockReferences = New Stack(Of BlockReference)(cumulativeStack)
            End If

            SearchNodeInTree(tv, blockReferences, selectedEntity)
        End Sub

        Sub SearchNodeInTree(ByVal tv As TreeView, ByVal blockReferences As Stack(Of BlockReference), ByVal selectedEntity As Model.SelectedItem, ByVal Optional parentTn As TreeNode = Nothing)
            If blockReferences.Count = 0 AndAlso selectedEntity Is Nothing Then Return
            Dim tnc As ItemCollection = tv.Items
            If parentTn IsNot Nothing Then tnc = parentTn.Items

            If blockReferences.Count > 0 Then
                Dim br As BlockReference = blockReferences.Pop()

                For Each tn As TreeNode In tnc

                    If ReferenceEquals(br, tn.Tag) Then

                        If blockReferences.Count > 0 Then
                            tn.IsExpanded = True
                            SearchNodeInTree(tv, blockReferences, selectedEntity, tn)
                        Else

                            If selectedEntity IsNot Nothing Then

                                For Each childNode As TreeNode In tn.Items

                                    If ReferenceEquals(selectedEntity.Item, childNode.Tag) Then
                                        If childNode.ParentNode IsNot Nothing Then childNode.ParentNode.IsExpanded = True
                                        childNode.IsSelected = True
                                        Exit For
                                    End If
                                Next
                            Else
                                tn.IsSelected = True
                            End If
                        End If

                        Return
                    End If
                Next
            Else

                If selectedEntity IsNot Nothing Then

                    For Each childNode As TreeNode In tnc

                        If ReferenceEquals(selectedEntity.Item, childNode.Tag) Then
                            childNode.IsSelected = True
                            Exit For
                        End If
                    Next
                End If
            End If
        End Sub

        Sub CollapseAll(ByVal tv As TreeView)
            For Each i As TreeNode In tv.Items
                i.IsExpanded = False
            Next
        End Sub
    End Module

    Public Class TreeNode
        Inherits FrameworkElement

        Public Sub New(ByVal parent As TreeNode)
            Items = New TreeView().Items
            ParentNode = parent
        End Sub

    Public Sub New(ByVal parent As TreeNode, ByVal _text As String, ByVal isPart As Boolean)
        Me.New(parent)
        Text = _text

        If isPart Then
            Icon = New BitmapImage(GetUriFromResource("part_icon.png"))
        Else
            Icon = New BitmapImage(GetUriFromResource("component_icon.png"))
        End If
    End Sub

    Public Property Text As String
        Public Property Icon As BitmapImage

        Private Shared Function GetUriFromResource(ByVal resourceFilename As String) As Uri
            Return New Uri("pack://application:,,,/Resources/" & resourceFilename)
        End Function

        Public Property ParentNode As TreeNode
        Public Property Items As ItemCollection
        Public Shared ReadOnly IsSelectedProperty As DependencyProperty = DependencyProperty.Register("IsSelected", GetType(Boolean), GetType(TreeNode), New PropertyMetadata(Nothing))

        Public Property IsSelected As Boolean
            Get
                Return CBool(GetValue(IsSelectedProperty))
            End Get
            Set(ByVal value As Boolean)
                SetValue(IsSelectedProperty, value)
            End Set
        End Property

        Public Shared ReadOnly IsExpandedProperty As DependencyProperty = DependencyProperty.Register("IsExpanded", GetType(Boolean), GetType(TreeNode), New PropertyMetadata(Nothing))

        Public Property IsExpanded As Boolean
            Get
                Return CBool(GetValue(IsExpandedProperty))
            End Get
            Set(ByVal value As Boolean)
                SetValue(IsExpandedProperty, value)
            End Set
        End Property

        Public Overrides Function ToString() As String
            Return Text
        End Function

        Public Sub Remove()
            While Items.Count > 0
                CType(Items(0), TreeNode).Remove()
            End While

            If ParentNode IsNot Nothing Then ParentNode.Items.Remove(Me)
        End Sub
End Class